client/network/src/protocol.rs

```diff
@@ -1454,6 +1433,63 @@ impl<B: BlockT, H: ExHashT> NetworkBehaviour for Protocol<B, H> {
 			return Poll::Ready(NetworkBehaviourAction::GenerateEvent(message));
 		}
 
+		// Check for finished outgoing requests.
+		let mut finished_block_requests = Vec::new();
+		for (id, peer) in self.context_data.peers.iter_mut() {
+			if let Peer { block_request: Some((_, pending_response)), .. } = peer {
+				match pending_response.poll_unpin(cx) {
+					Poll::Ready(Ok(Ok(resp))) => {
+						let (req, _) = peer.block_request.take().unwrap();
+
+						let protobuf_response = match crate::schema::v1::BlockResponse::decode(&resp[..]) {
+							Ok(proto) => proto,
+							Err(e) => {
+								trace!(target: "sync", "Failed to decode block request to peer {:?}: {:?}.", id, e);
+								self.peerset_handle.report_peer(id.clone(), rep::BAD_MESSAGE);
+								self.behaviour.disconnect_peer(id);
+								continue;
+							}
+						};
+
+						finished_block_requests.push((id.clone(), req, protobuf_response));
+					},
+					Poll::Ready(Ok(Err(e))) => {
+						peer.block_request.take();
+						trace!(target: "sync", "Block request to peer {:?} failed: {:?}.", id, e);
+
+						match e {
+							RequestFailure::UnknownProtocol => {
+								debug_assert!(false, "Block request protocol should always be known.");
+							}
+							RequestFailure::Network(OutboundFailure::Timeout) => {
+								self.peerset_handle.report_peer(id.clone(), rep::TIMEOUT);
```

Comment:
Previous implementation would disconnect on timeout. Is that reasonable?

Reply:
`rep::TIMEOUT` seems more appropriate to me.

Reply:
The state machine in `sync.rs` relies on the disconnect happening. The only way it knows that a request has failed is by reporting a disconnect.
Not disconnecting is not an option unless we change `sync.rs` as well (which I recommend against doing).

Reply:
@tomaka I have adjusted the behavior accordingly, would you mind taking another look?